src/coreclr/nativeaot/Runtime/arm/GcProbe.S

```diff
@@ -8,6 +8,100 @@
 #include "AsmOffsets.inc"
 
 .global RhpGcPoll2
+.global RhpThrowHwEx
+
+// See PUSH_COOP_PINVOKE_FRAME, this macro is very similar, but also saves return registers
+// and accepts the register bitmask
+// Call this macro first in the method (no further prolog instructions can be added after this).
+//
+//  threadReg     : register containing the Thread* (this will be preserved).
+//  trashReg      : register that can be trashed by this macro
+//  BITMASK       : value to initialize m_dwFlags field with (register or #constant)
+.macro PUSH_PROBE_FRAME threadReg, trashReg, BITMASK
+    // Define the method prolog, allocating enough stack space for the PInvokeTransitionFrame and saving
+    // incoming register values into it.
+    PROLOG_PUSH         "{r0-r1}"           // Save return registers
```

Comment:
Do we need to push down SP and store sp/fp first?

Reply:
I don't think it's necessary. Other archs don't do it either. They just construct the `PInvokeTransitionFrame` on the stack and make its IP/SP/FP point to the user code before the hijack. That way the `StackFrameIterator` on the GC thread begins unwinding from the correct location. On the hijack thread this is just pushed, then waits for end of GC, and pops it in the reverse order.

Reply:
I am not sure if Linux/ARM32 has a sufficient red zone. I can't find any good docs about that. 
Even if there is, perhaps better not rely on it as it might be OS-specific.

Reply:
I'm not sure I fully understand your concern here. The ARM/Thumb instruction set is not really well tailored for the `add sp, #imm` pattern and then accessing the stack by relative writes. That's why it's not done in prologs, or `PUSH_COOP_PINVOKE_FRAME`.

Reply:
Ah, I misread this as if we possibly use the stack beyond SP (that could cause problems with signal handlers, in theory). But we are not, so this should be ok.